### PR TITLE
G-API: Adding rescaling by ROI after gapi::ParseSSD

### PR DESCRIPTION
**Pull request by mpashchenkov**
Resumption: #21106
Added custom::ParseSSD (upscale in_roi versions) to infer_single_roi sample.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### Build commands:
```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.4.1:20.04
build_image:Custom Win=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.4.1

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by mpashchenkov** [comment]
@sivanov-work, VPL sample without changes now.

**Comment by sivanov-work** [comment]
@mpashchenkov  thanks
vpl changes would be applied in standalone PR for VPP preproc ROI support to get rid merge conflicts

**Review by sivanov-work — APPROVED** [review]

File: modules/gapi/samples/infer_single_roi.cpp [hunk 1]
```diff
@@ -164,7 +219,7 @@ int main(int argc, char *argv[])
                   << std::endl;
         cv::GOpaque<cv::Rect> in_roi;
         auto blob = cv::gapi::infer<custom::FaceDetector>(in_roi, in);
-        cv::GArray<cv::Rect> rcs = cv::gapi::parseSSD(blob, sz, 0.5f, true, true);
+        auto rcs = custom::ParseSSD::on(blob, in_roi, sz);
```
Comment (alalek) [inline modules/gapi/samples/infer_single_roi.cpp:222]: What is the purpose of creation of new one?
What is the difference between them?
Could we extend existing one with extra mode?

/cc @dmatveev We need your approval here.

Comment (mpashchenkov) [reply modules/gapi/samples/infer_single_roi.cpp:222]: @alalek, `custom::ParseSSD` upscales ssd result by input roi. Unfortunately `gapi::parseSSD` kernels don't have this functionality.

Comment (alalek) [reply modules/gapi/samples/infer_single_roi.cpp:222]: 3rd question is left.

Before implementing some functionality in samples only we need to evaluate its value to be a part of related module itself.

Comment (dmatveev) [reply modules/gapi/samples/infer_single_roi.cpp:222]: @alalek running a detector within ROI is mainly a synthetic case, so custom kernel is OK.

However, I'd consider using our "standard" parseSSD/parseYolo functions and introducing a custom kernel which makes just a coordinate shift based on the ROI. @mpashchenkov  could you do it that way?

Comment (mpashchenkov) [reply modules/gapi/samples/infer_single_roi.cpp:222]: Yes, it will be a simple overload as far as I remember.

**Comment by asmorkalov** [comment]
@dmatveev Will you have a chance to review th PR in mean time? The PR fixes some old bug and it'll be great to merge it for 4.7.

**Comment by asmorkalov** [comment]
@sivanov-work @dmatveev friendly reminder.

File: modules/gapi/samples/onevpl_infer_single_roi.cpp [hunk 2]
```diff
@@ -207,51 +207,16 @@ GAPI_OCV_KERNEL(OCVBBoxes, BBoxes) {
...
-            rc.height = static_cast<int>(rc_bottom * up_roi.height) - rc.y;
-            rc.x += in_roi.x;
-            rc.y += in_roi.y;
-            out_objects.emplace_back(rc & surface);
+GAPI_OCV_KERNEL(OCVRescaleByROI, RescaleByROI) {
+    static void run(const std::vector<cv::Rect>& in_objects,
+                    const cv::Rect& in_roi,
+                    const cv::Size& in_size,
+                    std::vector<cv::Rect>& out_objects) {
+        for (auto&& rc : in_objects) {
```
Comment (SergeyIvanov87) [inline modules/gapi/samples/onevpl_infer_single_roi.cpp:215]: there was ' out_objects.clear()' in previous version. Do we need clear it here too?
it yes, then additionally we might use ' out_objects.reserve(in_objects.size())` as well

File: modules/gapi/samples/infer_single_roi.cpp [hunk 3]
```diff
@@ -164,7 +186,8 @@ int main(int argc, char *argv[])
                   << std::endl;
         cv::GOpaque<cv::Rect> in_roi;
         auto blob = cv::gapi::infer<custom::FaceDetector>(in_roi, in);
-        cv::GArray<cv::Rect> rcs = cv::gapi::parseSSD(blob, sz, 0.5f, true, true);
+        auto temp_rcs = cv::gapi::parseSSD(blob, sz, 0.5f, false, true);
```
Comment (SergeyIvanov87) [inline modules/gapi/samples/infer_single_roi.cpp:189]: what  does changing 'true'->'false' mean?

File: modules/gapi/samples/infer_single_roi.cpp [hunk 4]
```diff
@@ -124,6 +130,21 @@ GAPI_OCV_KERNEL(OCVBBoxes, BBoxes) {
     }
 };
 
+
+GAPI_OCV_KERNEL(OCVRescaleByROI, RescaleByROI) {
```
Comment (dmatveev) [inline modules/gapi/samples/infer_single_roi.cpp:134]: What is the reason to rescale objects?

Don't we need to shift them (e.g. change x/y coordinates only) when projecting a detection from ROI to the parent image?

File: modules/gapi/samples/onevpl_infer_single_roi.cpp (hunk 2 above)
Comment (dmatveev) [reply modules/gapi/samples/onevpl_infer_single_roi.cpp:215]: +1

**Comment by mpashchenkov** [comment]
I'll be back later. I don't have time for this right now.

**Pull request closed without merging**